src/python/pants/bin/remote_pants_runner.py

```diff
@@ -175,44 +127,16 @@ def _connect_and_execute(self, pantsd_handle: PantsDaemonClient.Handle) -> ExitC
             ),
         }
 
-        # Instantiate a NailgunClient.
-        client = NailgunClient(
-            port=port,
-            remote_pid=pid,
-            ins=sys.stdin,
-            out=sys.stdout.buffer,
-            err=sys.stderr.buffer,
-            exit_on_broken_pipe=True,
-            metadata_base_dir=pantsd_handle.metadata_base_dir,
-        )
-
-        timeout = global_options.pantsd_pailgun_quit_timeout
-        pantsd_signal_handler = PailgunClientSignalHandler(client, pid=pid, timeout=timeout)
-        with ExceptionSink.trapped_signals(pantsd_signal_handler), STTYSettings.preserved():
-            # Execute the command on the pailgun.
-            return client.execute(self._args[0], self._args[1:], modified_env)
+        command = self._args[0]
+        args = self._args[1:]
 
-    def _extract_remote_exception(self, pantsd_pid, nailgun_error):
-        """Given a NailgunError, returns a Terminated exception with additional info (where
-        possible).
+        def signal_fn() -> bool:
+            return ExceptionSink.signal_sent() is not None
 
-        This method will include the entire exception log for either the `pid` in the NailgunError,
-        or failing that, the `pid` of the pantsd instance.
-        """
-        sources = [pantsd_pid]
-
-        exception_text = None
-        for source in sources:
-            log_path = ExceptionSink.exceptions_log_path(for_pid=source)
-            exception_text = maybe_read_file(log_path)
-            if exception_text:
-                break
-
-        exception_suffix = (
-            "\nRemote exception:\n{}".format(exception_text) if exception_text else ""
-        )
-        return self.Terminated(
-            "abruptly lost active connection to pantsd runner: {!r}{}".format(
-                nailgun_error, exception_suffix
+        rust_nailgun_client = native.new_nailgun_client(port=port)
+        pantsd_signal_handler = PailgunClientSignalHandler(pid=pid)
+
+        with ExceptionSink.trapped_signals(pantsd_signal_handler), STTYSettings.preserved():
```

Comment:
In the context of #10792, it seems that the signal handling aspect across a nailgun connection is definitely something we would like to move into rust. With #10057, it seems we likely want to put all the tty handling into rust too (eventually!).

Would it be reasonable to leave:
```suggestion
        # TODO(#10792) move ExceptionSink to rust
        # TODO(#10057) move tty handling to rust
        with ExceptionSink.trapped_signals(pantsd_signal_handler), STTYSettings.preserved():
```

Reply:
Ideally we won't be doing signal handling across the nailgun connection at all, once this commit is merged and we start using the `heartbeat` faculty of the `nails` crate. I'm trying to keep this PR scoped as narrowly as possible to just replacing the existing python nailgun client with this rust client. It might be a good idea to move `ExceptionSink` to rust, but I think it's outside teh scope of this PR one way or the other.